### PR TITLE
kv: refactor UnionStore

### PR DESCRIPTION
**Pull request by disksing**
1. Export kv.UnionStore as an interface.
2. Introduce `Retriever` and `Updater` interfaces.
3. Move common parts of localstore/txn and hbase/txn into kv.UnionStore.

@ngaut

### CODE REVIEW

**Comment by ngaut** [comment]
LGTM @coocood @siddontang

File: kv/union_store.go [hunk 1]
```diff
@@ -112,29 +118,74 @@ func (us *UnionStore) Get(key []byte) (value []byte, err error) {
...
 }
 
-// Set implements the Store Set interface.
-func (us *UnionStore) Set(key []byte, value []byte) error {
-	return us.WBuffer.Set(key, value)
+// Set implements the MemStore interface.
+func (us *unionStore) Set(key Key, value []byte) error {
+	if len(value) == 0 {
+		// Incase someone use it in the wrong way, we can figure it out immediately.
+		debug.PrintStack()
```
Comment (qiuyesuifeng) [inline kv/union_store.go]: Can this line be removed?

File: kv/kv.go [hunk 2]
```diff
@@ -119,26 +119,51 @@ const (
 	PresumeKeyNotExists
 )
 
-// Transaction defines the interface for operations inside a Transaction.
-// This is not thread safe.
-type Transaction interface {
-	// Get gets the value for key k from KV store.
+// MemStore is an abstract KV storage.
+type MemStore interface {
```
Comment (coocood) [inline kv/kv.go]: Why not name it KVStore?
It is an interface, implementation can be pesistent store.

Comment (disksing) [inline kv/kv.go]: yeah, MemStore may be not a good choice. I thought of KVStore in the first place, but `kv.KVStore` seems verbose. Another choice is using just `Store`, but we have already got a `Store`. Call for some discussions. @qiuyesuifeng @siddontang

File: kv/union_store.go (hunk 1 above)
Comment (disksing) [inline kv/union_store.go]: Maybe the best way to figure it out immediately is using panic instead?

**Comment by disksing** [comment]
@ngaut @siddontang @qiuyesuifeng PTAL.

**Comment by coocood** [comment]
Maybe `Mutator` is better than `Updater`.

File: kv/union_store.go [hunk 3]
```diff
@@ -16,6 +16,8 @@ package kv
 import (
 	"bytes"
 
+	"strconv"
```
Comment (siddontang) [inline kv/union_store.go]: remove above space.

**Comment by coocood** [comment]
LGTM

**Comment by ngaut** [comment]
LGTM

**Pull request merged**